README.md
# MedCEP
<img src="https://github.com/vinnysoft/MedCEP/blob/master/Addons/web/naviox/images/nemo.jpg" alt="Logo NEMO" width="140px" height="120px"/> <img src="https://github.com/vinnysoft/MedCEP/blob/master/Addons/web/naviox/images/ufes.png" alt="Logo UFES" width="140px" height="120px"/>

Ferramenta para Medição de Software e Controle Estatístico de Processos. 
<br/>Baseada na arquitetura de referência de medição de software de MARETTO (2014).

##Requisitos
- Java SDK 7 Update 75 (x64) - [Download] (http://www.oracle.com/technetwork/pt/java/javase/downloads/jdk7-downloads-1880260.html)
- IDE Eclipse Luna for EE Developers (x64) SR2 - [Download] (https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/luna/SR2/eclipse-jee-luna-SR2-win32-x86_64.zip)
- PostgreSQL 9.4.1 (x64) - [Download] (http://www.enterprisedb.com/products-services-training/pgdownload)
- Apache Tomcat 7.0.59 (x64) - [Download] (http://mirrors.koehn.com/apache/tomcat/tomcat-7/v7.0.59/bin/apache-tomcat-7.0.59-windows-x64.zip)

##Instruções para configuração do ambiente
1. Instalar o Java SDK.
2. Instalar o PostgreSQL com usuário e senha <b>"postgres"</b>.
3. Criar um banco de dados no PostgreSQL chamado <b>"medcep"</b> e informar como dono o usuário <b>"postgres"</b>.
4. Descompactar o Apache Tomcat em C:\apache-tomcat-7.0.59 (caminho padrão). Caso queira executar o Apache Tomcat em outro diretório, atualizar a variável <b>tomcat.dir</b> do arquivo <b>MedCEP/properties/openxava.properties</b>. 
5. Descompactar o IDE Eclipse Luna.
6. Executar o Eclipse.
7. Clonar o repositório MedCEP do GitHub.
8. Configurar o Apache Tomcat como servidor web (<b>Window -> Preferences -> Server -> Runtime Environments -> Add...</b>). 
9. Na aba Server (parte inferior do Eclipse), dar duplo clique em <b>Tomcat v7 Server</b>. No arquivo que abrir, selecionar a opção <b>"Use Tomcat Installation (takes control of Tomcat installation)"</b> na área Server Locations. 
10. Iniciar o Apache Tomcat.
11. Abrir a pasta do projeto MedCEP no IDE Eclipse.
12. Procurar pelo arquivo <b>build.xml</b>.
13. Clicar com o botão direito do mouse, selecionar <b>"Run as"</b> e escolher a 3º opção <b>"Ant Build..."</b>.
14. Defina um nome, por exemplo <b>"MedCEP.Implantar"</b>, depois clique em <b>Apply</b>.
15. Clique em <b>Run</b>.
16. Executar a Ant Build <b>MedCEP.Implantar</b>.
17. Abrir a aplicação no browser pela URL: [http://localhost:8080/MedCEP](http://localhost:8080/MedCEP).

Obs.: Caso, após a instrução 17, não tenha sido possível abrir a aplicação através da URL via browser, faça:

1. Pare o Apache Tomcat.
2. Inicie novamente o Apache Tomcat.
3. Execute a Ant Build <b>MedCEP.Implantar</b>.
4. Abra a aplicação no browser pela URL: [http://localhost:8080/MedCEP](http://localhost:8080/MedCEP).

##Configuração do Mantis Bug Tracking
###Requisitos
- Wamp Server 2.5 (Windows + Apache + MySQL + PHP server) (x64) - [Download](http://sourceforge.net/projects/wampserver/files/WampServer%202/Wampserver%202.5/wampserver2.5-Apache-2.4.9-Mysql-5.6.17-php5.5.12-64b.exe/download)
- Mantis Bug Tracking 1.2.19 - [Download] (http://sourceforge.net/projects/mantisbt/files/mantis-stable/1.2.19/mantisbt-1.2.19.zip/download)
 
###Instruções para configuração
1. Instalar o Wamp Server 2.5
2. Alterar o idioma para Português (clicar com botão direito do mouse no ícone do Wamp Server 2.5 na bandeja do Windows, depois em <b>"Language -> portuguese"</b>).
2. Abrir o phpMyAdmin (clicar no ícone do Wamp Server 2.5 na bandeja do Windows, depois em <b>"phpMyAdmin"</b>).
3. Clicar em <b>Utilizadores</b> no menu superior.
4. Clicar em <b>Alterar Privilégios</b> para o utilizador <b>root</b> na máquina <b>localhost</b>.
5. No grupo <b>Alterar a palavra-passe</b>, informar a senha root e clicar em <b>Executar</b>.
3. Clicar em <b>New</b> no menu lateral esquerdo, para criar o banco de dados para o Mantis Bug Tracking.
4. Informar o nome <b>"mantis"</b> e clicar em <b>Criar</b>.
5. O banco de dados <b>"mantis"</b> aparecerá na listagem logo abaixo (na mesma tela). Clicar em <b>Verificar Privilégios</b>.
6. Clicar em <b>Adicionar utilizador</b>.
7. Informar User name: mantis, Host: localhost, Palavra-passe: mantis, Re-type: mantis, clicar em <b>"Executar"</b>.
2. Abrir o diretório <b>"www"</b> (clicar no ícone do Wamp Server 2.5 na bandeja do Windows, depois em <b>"diretório www"</b>).
3. Descompactar o arquivo zip do Mantis Bug Tracking no <b>"diretório www"</b>.
4. Renomear a pasta do Mantis Bug Tracking de <b>"mantisbt-1.2.19"</b> para <b>"mantis"</b>.
5. Acessar a URL: [http://localhost/mantis/](http://localhost/mantis/)
6. Informar <b>Type of Database: MySQL (default), Hostname: localhost, Username: mantis, Password: mantis, Database name: mantis</b>, clicar em <b>"Install/Upgrade Database"</b>.
7. Após a instalação, acessar novamente a URL [http://localhost/mantis/](http://localhost/mantis/)
8. Informar o usuário <b>Administrator</b> e senha <b>root</b> para acessar o Mantis Bug Tracking.

<b><i>IMPORTANTE: Os passos informados nas instruções de configuração foram baseados em um ambiente de desenvolvimento. Para ambientes de produção, as variáveis como senha, URL e local de instalação deverão ser alteradas de acordo com cada necessidade.</i></b>

## Referência
C. X. MARETTO and M. P. BARCELLOS, <b>“Uma Arquitetura de Referência para Medição de Software,”</b> in XIII Simpósio Brasileiro de Qualidade de Software (SBQS 2014), 2014.


OpenXava/changelog.txt
Change Log for OpenXava project
===============================

OpenXava 5.2.1 (2015-3-2)
-------------------------
- Methods showDialog(), closeDialog(), showNewView() and returnToPreviousView() available from OnChangePropertyBaseAction.
- New removeBlanks() method in Strings utility class.
- Entity from View has nulls for non-existing references, before they were empty objects.
- Fix: Reference with calculated property in an entity is cleared after modifying a collection in the same entity.
- Fix: View.addValues() does not keep original values in subviews. 
- Fix: Sort and filter lost in collections after editing/viewing an element with a dialog.
- Fix: ArrayIndexOutOfBoundsException trace in tab with calculated properties.
- Fix: Misleading message "Trying to load controller twice" in the log. 
- Fix: Apostrophe used in labels files is not shown in tooltips. 
- Fix: Several except-for-tab in tabs-default-values fails.
- Fix: Calling getView().getEntity() when there is an element collection in the view fails. 
- Fix: helpPrefix from xava.properties does not work for external URLs.
- Fix: It is possible to add programmatically the same action several times. 
- Fix: View.getEntity() fails when there is an empty reference in the UI. 
- Fix: Dialog is not shown from an @OnChange action. 
- Fix: Dangerous JavaScript content possible in list mode by a XSS attack.
- Fix: An IForwardAction from an IChainAction with on-init=true does not do the forward. 
- Fix: Collection customization is not stored permanently if @ListProperties is used. 
- Fix: DWR signature warning in log.
- Fix: View.getEntity() when creating does not load the references so combo descriptions are not available.
- Fix: FILES stereotype fails from embedded reference.
- Fix: XSS protection does not include iframe and window.location.
- Fix: Error "i.contentWindow is null" on closing a dialog with HTML_TEXT stereotype.
- Fix: Frames for groups with blanks in their name do not open and close. 
- Fix: Section and group names with blanks produces some misbehavior.
- Fix: A reference to parent cannot be changed in a composite structure.

OpenXava 5.2 (2015-1-26)
------------------------
- Moving columns in list with drag and drop.
- Removing a column from a list is done instantly and with a fade effect.
- The list for adding columns to a list is shown inside a dialog.
- The list for adding columns to a list no longer uses pagination, but a scrollbar.
- The list for adding columns to a list shows only until second level properties at first, and all properties on demand.
- The list for adding columns to a list is ordered by the localized column labels of the properties.
- The list for adding columns to a list no longer shows the internal properties names.
- Fade effect to show/hide the filter part of the list and the customization icons.
- Using JavaScript to show/hide the list customization icons, without an AJAX request.
- List customization icons are hidden automatically after executing any action.
- 20% performance improvement in automated module tests.
- HtmlUnit upgraded to 2.15.
- jQuery UI upgraded to 1.11.2.
- jQuery upgraded to 1.11.2.  
- Upgraded Commons Lang3 library to 3.3.2. 
- New move() method in XCollections utility class.
- The browser property in xava-junit.properties is no longer supported.
- New createWar ant task included by default in new projects.
- Frame spacing when several frames in a row adjusted. 
- Performance improvement for label resolution.
- Fix: After removing a column from a list the filter data is gone.
- Fix: Test code does not work after calling ModuleTestBase.logout().
 
OpenXava 5.1.1 (2014-11-24)
---------------------------
- New isNumeric() method in Strings utility class.
- Some new common use labels to be used in applications.
- Fix: It does not work with WebSphere Portal (broken since v5.0).
- Fix: Required icon is not shown for descriptions lists. 
- Fix: Too much space between UI elements when a descriptionLists is located after the first column.
- Fix: Accessing to an application from a mobile device shows a resource not found page.

OpenXava/generator/PropertyPG.java

/* Generated by TL2Java Version 1.2, April 9, 2001 */
import org.w3c.dom.*;
import java.io.*;
import java.util.*;
import org.openxava.util.Strings;
import org.openxava.util.XavaException;
import org.openxava.model.meta.*;
import org.openxava.generators.*;
import org.openxava.calculators.*;
import org.openxava.util.meta.MetaSet;
import org.openxava.mapping.*;

/**
 * Program Generator created by TL2Java
 * @version Fri Mar 21 14:15:30 CET 2014
 */
public class PropertyPG {
    Properties properties = new Properties();


private String getException() {
	return ejb?"EJBException":"RuntimeException";
}

private MetaProperty property=null;
public void setProperty(MetaProperty property) {
	this.property = property;
}

private boolean ejb=false;
public void setEjb(boolean ejb) {
	this.ejb = ejb;
}

private MetaModel metaModel=null;
private ModelMapping modelMapping=null;
public void setMetaModel(MetaModel metaModel) throws XavaException {
	this.metaModel = metaModel;
	if (metaModel instanceof MetaAggregateForReference) {
		this.modelMapping = metaModel.getMetaComponent().getEntityMapping();
	}
	else {
		this.modelMapping = metaModel.getMapping();
	}	
}

private boolean attributeIsPublic = false;
private String attributeAccessLevel = "private";
public void setAttributePublic(boolean attributeIsPublic) {
	this.attributeIsPublic = attributeIsPublic;
	this.attributeAccessLevel = attributeIsPublic?"public":"private";
}

private boolean isPublic = true;
private String propertyAccessLevel = "public";
public void setPublic(boolean isPublic) {
	this.isPublic = isPublic;
	this.propertyAccessLevel = isPublic?"public":"private";
}
public static void generate(XPathContext context, ProgramWriter out, MetaProperty property) throws XavaException {
	generate(context, out, property, true, false);
}

public static void generateEJB(XPathContext context, ProgramWriter out, MetaProperty property) throws XavaException {
	generate(context, out, property, true, true);
}

public static void generatePrivate(XPathContext context, ProgramWriter out, MetaProperty property, boolean ejb) throws XavaException {
	generate(context, out, property, false, ejb);
}

private static void generate(XPathContext context, ProgramWriter out, MetaProperty property, boolean isPublic, boolean isEjb) throws XavaException {
	PropertyPG pg = new PropertyPG();
	pg.setProperty(property);
	pg.setMetaModel(property.getMetaModel());
	pg.setPublic(isPublic);
	pg.setEjb(isEjb);
	pg.generate(context, out);
}




    /**
     * This method generates the output given a context and output stream
     */
    public boolean generate(XPathContext context, ProgramWriter out) {
        try {    
    	if (GeneratorFactory.has(property, ejb)) {
    		IPropertyCodeGenerator generator = GeneratorFactory.create(property, ejb);
    		String propertyCode = generator.generate();
    
    out.print(" \n\t");
    out.print(propertyCode);
    		
    		return true;
    	}
    	String propertyName = Strings.firstUpper(property.getName());	
    	String type = property.getTypeName();	
    	String get = type.equals("boolean")?"is":"get";	
    	String tagValueObject = ejb && isPublic?"@ejb.value-object match=\"persistentCalculatedAndAggregate\"":"";
    	String tagInterfaceMethod = ejb && isPublic?"@ejb:interface-method":"";
    	if (property.isCalculated()) {
    		// Calculated ones
    	
    out.print(" \t\n\t/**\n\t * ");
    out.print(tagValueObject);
    out.print("\n\t * ");
    out.print(tagInterfaceMethod);
    out.print("\n\t */\n\tpublic ");
    out.print(type);
    out.print(" ");
    out.print(get);
    out.print(propertyName);
    out.print("() {");
    
    		if (ejb) {
    
    out.print(" \n\t\tboolean cmtActivated = false;\n\t\tif (!org.openxava.hibernate.XHibernate.isCmt()) {\n\t\t\torg.openxava.hibernate.XHibernate.setCmt(true);\n\t\t\tcmtActivated = true;\n\t\t}");
    	
    		} 
    			MetaCalculator calculator = property.getMetaCalculator();
    			String calculatorClass = calculator.getClassName();
    			String qualifiedPropertyName = Strings.change(property.getName(), "_", "."); // for aggregate member case
    			
    out.print(" \t\t\n\t\ttry {\t\t\t\n\t\t\t");
    out.print(calculatorClass);
    out.print(" ");
    out.print(property.getName());
    out.print("Calculator= (");
    out.print(calculatorClass);
    out.print(")\n\t\t\t\tgetMetaModel().getMetaProperty(\"");
    out.print(qualifiedPropertyName);
    out.print("\").getMetaCalculator().createCalculator();");
    	
    			Iterator itSets = calculator.getMetaSetsWithoutValue().iterator();
    			while (itSets.hasNext()) {
    				MetaSet set = (MetaSet) itSets.next();
    				String propertyNameInCalculator = Strings.firstUpper(set.getPropertyName());
    				String propertyNameFrom = set.getPropertyNameFrom();
    				MetaProperty p = null;
    				try {
    					p = metaModel.getMetaProperty(propertyNameFrom);
    				}
    				catch (org.openxava.util.ElementNotFoundException ex) {
    					// Trying if it's referencing to its parent
    					String parentPrefix = Strings.firstLower(metaModel.getContainerModelName()) + ".";
    					if (propertyNameFrom.startsWith(parentPrefix)) {
    						String propertyInParent = propertyNameFrom.substring(parentPrefix.length());
    						p = metaModel.getMetaModelContainer().getMetaProperty(propertyInParent);
    					}
    					else throw ex;
    				}
    				if (propertyNameFrom.indexOf('.') >= 0) {
    					StringTokenizer st = new StringTokenizer(propertyNameFrom, ".");
    					boolean moreThan2Levels = st.countTokens() > 2;
    					StringBuffer propertyNameFromInJava = new StringBuffer();
    					boolean isEmbededKey = ejb && ((p.isKey() && !moreThan2Levels) || (p.getMetaModel() instanceof MetaAggregate));
    					while (st.hasMoreTokens()) {
    						String token = st.nextToken();
    						if (propertyNameFromInJava.length() > 0) {
    							if (isEmbededKey && !st.hasMoreTokens()) {
    								propertyNameFromInJava.append("_");
    							}
    							else {
    								if (isEmbededKey) {
    									propertyNameFromInJava.append("Remote");
    								}	
    								propertyNameFromInJava.append("().get");						
    							}						
    						}
    						if (isEmbededKey && !st.hasMoreTokens()) {
    							propertyNameFromInJava.append(token);
    						}
    						else {
    							propertyNameFromInJava.append(Strings.firstUpper(token));	
    						}
    					}			
    					propertyNameFrom = propertyNameFromInJava.toString();
    				}
    				else {
    					propertyNameFrom = Strings.firstUpper(propertyNameFrom);
    				}
    				int idx = propertyName.indexOf('_');
    				if (idx >= 0) {
    					String aggregateName = propertyName.substring(0, idx + 1);
    					propertyNameFrom = aggregateName + Strings.firstLower(propertyNameFrom);
    				}
    				String getPropertyFrom = "boolean".equals(p.getTypeName())?"is":"get";
    				String value = set.getValue();
    				if (set.hasValue()) {
    			
    out.print(" \n\t\t\t");
    out.print(property.getName());
    out.print("Calculator.set");
    out.print(propertyNameInCalculator);
    out.print("(\"");
    out.print(value);
    out.print("\");");
    
    				} else {	
    			
    out.print("  \t\n\t\t\t\n\t\t\t");
    out.print(property.getName());
    out.print("Calculator.set");
    out.print(propertyNameInCalculator);
    out.print("(");
    out.print(getPropertyFrom);
    out.print(propertyNameFrom);
    out.print("());");
    	}} // else/sets 	 
    			if (IModelCalculator.class.isAssignableFrom(Class.forName(calculatorClass))) { 
    			
    out.print(" \n\t\t\t\t");
    out.print(property.getName());
    out.print("Calculator.setModel(this);");
    } 
    			if (IEntityCalculator.class.isAssignableFrom(Class.forName(calculatorClass))) { 
    			
    out.print(" \n\t\t\t\t");
    out.print(property.getName());
    out.print("Calculator.setEntity(this);");
    } 
    			if (IJDBCCalculator.class.isAssignableFrom(Class.forName(calculatorClass))) {
    				if (ejb) { 
    			
    out.print(" \n\t\t\t\t");
    out.print(property.getName());
    out.print("Calculator.setConnectionProvider(getPortableContext());");
    			
    				}
    				else { // not ejb
    			
    out.print(" \n\t\t\t\t");
    out.print(property.getName());
    out.print("Calculator.setConnectionProvider(DataSourceConnectionProvider.getByComponent(\"");
    out.print(metaModel.getMetaComponent().getName());
    out.print("\"));");
    			
    				}
    			}  
    			String calculateValueSentence = property.getName() + "Calculator.calculate()";		
    			
    out.print(" \n\t\t\treturn ");
    out.print(Generators.generateCast(property.getTypeName(), calculateValueSentence));
    out.print(";\n\t\t}\n\t\tcatch (NullPointerException ex) {\n\t\t\t// Usually for multilevel property access with null references");
    if (boolean.class.equals(property.getType())) { 
    out.print("\n\t\t\treturn false;");
    } else if (property.getType().isPrimitive()) { 
    out.print(" \n\t\t\treturn 0;");
    } else { 
    out.print(" \n\t\t\treturn null;");
    } 
    out.print(" \t\t\t\n\t\t}\n\t\tcatch (Exception ex) {\n\t\t\tex.printStackTrace();\n\t\t\tthrow new ");
    out.print(getException());
    out.print("(XavaResources.getString(\"generator.calculate_value_error\", \"");
    out.print(propertyName);
    out.print("\", \"");
    out.print(metaModel.getName());
    out.print("\", ex.getLocalizedMessage()));\n\t\t}");
    
    		if (ejb) {
    
    out.print(" \n\t\tfinally {\n\t\t\tif (cmtActivated) {\n\t\t\t\torg.openxava.hibernate.XHibernate.setCmt(false);\n\t\t\t}\n\t\t}");
    	
    		} 
    
    out.print(" \t\t\n\t}\n\tpublic void set");
    out.print(propertyName);
    out.print("(");
    out.print(type);
    out.print(" new");
    out.print(propertyName);
    out.print(") {\n\t\t// for it is in value object\n\t\t// para que aparezca en los value objects\n\t}");
    
    	}
    	else {
    		// Not calculateds
    		String column = modelMapping.getColumn(property.getName());			
    		PropertyMapping propertyMapping = modelMapping.getPropertyMapping(property.getName());
    		
    		String ejbPkField = "";
    		String ejbSetInterfaceMethod = ejb && isPublic?"@ejb:interface-method":"";
    		if (property.isKey()) {
    			ejbPkField = "@ejb:pk-field";
    			ejbSetInterfaceMethod = "";			
    		}
    		if (propertyMapping.hasConverter() || propertyMapping.hasMultipleConverter()) { // Common for multiple and normal converter
    			String converterClass = propertyMapping.hasConverter()?"org.openxava.converters.IConverter":propertyMapping.getMultipleConverterClassName();		
    			String multiple = propertyMapping.hasMultipleConverter()?"Multiple":"";
    			
    out.print(" \n\tprivate static ");
    out.print(converterClass);
    out.print(" ");
    out.print(property.getName());
    out.print("Converter;\n\tprivate ");
    out.print(converterClass);
    out.print(" get");
    out.print(propertyName);
    out.print("Converter() {\n\t\tif (");
    out.print(property.getName());
    out.print("Converter == null) {\n\t\t\ttry {\n\t\t\t\t");
    out.print(property.getName());
    out.print("Converter = (");
    out.print(converterClass);
    out.print(") \n\t\t\t\t\tgetMetaModel().getMapping().get");
    out.print(multiple);
    out.print("Converter(\"");
    out.print(property.getName());
    out.print("\");\n\t\t\t}\n\t\t\tcatch (Exception ex) {\n\t\t\t\tex.printStackTrace();\n\t\t\t\tthrow new ");
    out.print(getException());
    out.print("(XavaResources.getString(\"generator.create_converter_error\", \"");
    out.print(property.getName());
    out.print("\"));\n\t\t\t}\n\t\t\t\n\t\t}\t\n\t\treturn ");
    out.print(property.getName());
    out.print("Converter;\n\t}");
    			
    		}
    		if (propertyMapping.hasMultipleConverter()) { // With converter for multiple fields
    			Iterator itCmpFields = propertyMapping.getCmpFields().iterator();
    			while (itCmpFields.hasNext()) {
    				CmpField cmpField = (CmpField) itCmpFields.next();
    				if (ejb) {
    		
    out.print(" \n\t/**\t \n\t * @ejb:persistent-field\n\t * @jboss:column-name \"");
    out.print(cmpField.getColumn());
    out.print("\"\n\t */\t\t\n\tpublic abstract ");
    out.print(cmpField.getCmpTypeName());
    out.print(" get");
    out.print(propertyName);
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print("();\n\tpublic abstract void set");
    out.print(propertyName);
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print("(");
    out.print(cmpField.getCmpTypeName());
    out.print(" newValue);");
    
    				}
    				else { // not ejb
    		
    out.print(" \n\tprivate ");
    out.print(cmpField.getCmpTypeName());
    out.print(" ");
    out.print(property.getName());
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print(";\n\tprivate ");
    out.print(cmpField.getCmpTypeName());
    out.print(" get");
    out.print(propertyName);
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print("() {\n\t\treturn ");
    out.print(property.getName());
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print(";\n\t}\n\tprivate void set");
    out.print(propertyName);
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print("(");
    out.print(cmpField.getCmpTypeName());
    out.print(" newValue) {\n\t\tthis.");
    out.print(property.getName());
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print(" = newValue;\n\t}");
    	
    				}
    			}
    		
    out.print(" \n\t/**\n\t * ");
    out.print(tagValueObject);
    out.print("\n\t * ");
    out.print(tagInterfaceMethod);
    out.print("\n\t */\n\tpublic ");
    out.print(type);
    out.print(" ");
    out.print(get);
    out.print(propertyName);
    out.print("() {\n\t\ttry {");
    
    			itCmpFields = propertyMapping.getCmpFields().iterator();
    			while (itCmpFields.hasNext()) {
    				CmpField cmpField = (CmpField) itCmpFields.next();				
    		
    out.print(" \n\t\t\tget");
    out.print(propertyName);
    out.print("Converter().set");
    out.print(Strings.firstUpper(cmpField.getConverterPropertyName()));
    out.print("(get");
    out.print(propertyName);
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print("());");
    		
    			}
    			String toJavaSentence = "get" + propertyName + "Converter().toJava()";
    		
    out.print(" \n\t\t\treturn ");
    out.print(Generators.generateCast(type, toJavaSentence));
    out.print(";\n\t\t}\n\t\tcatch (org.openxava.converters.ConversionException ex) {\n\t\t\tex.printStackTrace();\n\t\t\tthrow new ");
    out.print(getException());
    out.print("(XavaResources.getString(\"generator.conversion_error\", \"");
    out.print(propertyName);
    out.print("\", \"");
    out.print(metaModel.getName());
    out.print("\", \"");
    out.print(type);
    out.print("\"));\n\t\t}\n\t}\n\n\t/**\n\t * ");
    out.print(ejbSetInterfaceMethod);
    out.print("\n\t */\n\tpublic void set");
    out.print(propertyName);
    out.print("(");
    out.print(type);
    out.print(" new");
    out.print(propertyName);
    out.print(") {\n\t\ttry {");
    if (ejb) { 
    out.print("\n\t\t\tthis.modified = true;");
    } 
    
    			String argv = Generators.generatePrimitiveWrapper(type, "new" + propertyName);
    			
    out.print(" \n\t\t\tget");
    out.print(propertyName);
    out.print("Converter().toDB(");
    out.print(argv);
    out.print(");");
    
    			itCmpFields = propertyMapping.getCmpFields().iterator();
    			while (itCmpFields.hasNext()) {
    				CmpField cmpField = (CmpField) itCmpFields.next();				
    		
    out.print(" \n\t\t\tset");
    out.print(propertyName);
    out.print("_");
    out.print(cmpField.getConverterPropertyName());
    out.print("(get");
    out.print(propertyName);
    out.print("Converter().get");
    out.print(Strings.firstUpper(cmpField.getConverterPropertyName()));
    out.print("());");
    		
    			}
    		
    out.print(" \t\t\t\n\t\t}\n\t\tcatch (org.openxava.converters.ConversionException ex) {\n\t\t\tex.printStackTrace();\n\t\t\tthrow new ");
    out.print(getException());
    out.print("(XavaResources.getString(\"generator.conversion_db_error\", \"");
    out.print(propertyName);
    out.print("\", \"");
    out.print(metaModel.getName());
    out.print("\"));\n\t\t}\t\t\n\t}");
    
    		}
    		else if (propertyMapping.hasConverter()) { // Wint single field converter
    			String cmpType = propertyMapping.getCmpTypeName();
    			String getSentence = "get" + propertyName + "Converter().toJava(get_"+propertyName+"())";			
    			String setSentence = "get" + propertyName + "Converter().toDB(" + Generators.generatePrimitiveWrapper(type, "new" + propertyName) + ")";
    			if (!ejb) {
    
    out.print(" \n\t");
    out.print(attributeAccessLevel);
    out.print(" ");
    out.print(cmpType);
    out.print(" ");
    out.print(property.getName());
    out.print(";\n\tprivate ");
    out.print(cmpType);
    out.print(" get_");
    out.print(propertyName);
    out.print("() {\n\t\treturn ");
    out.print(property.getName());
    out.print(";\n\t}\n\tprivate void set_");
    out.print(propertyName);
    out.print("(");
    out.print(cmpType);
    out.print(" new");
    out.print(propertyName);
    out.print(") {\n\t\tthis.");
    out.print(property.getName());
    out.print(" = new");
    out.print(propertyName);
    out.print(";\n\t}");
    			
    			}
    			else {
    
    out.print(" \n\t/**\t \n\t * @ejb:persistent-field\n\t * ");
    out.print(ejbPkField);
    out.print("\n\t * @jboss:column-name \"");
    out.print(column);
    out.print("\"\n\t */\n\tpublic abstract ");
    out.print(cmpType);
    out.print(" get_");
    out.print(propertyName);
    out.print("();\n\tpublic abstract void set_");
    out.print(propertyName);
    out.print("(");
    out.print(cmpType);
    out.print(" new");
    out.print(propertyName);
    out.print(");");
    
    			}
    
    out.print(" \t\n\t\n\t/**\n\t * ");
    out.print(tagValueObject);
    out.print("\n\t * ");
    out.print(tagInterfaceMethod);
    out.print("\n\t */\n\t");
    out.print(propertyAccessLevel);
    out.print(" ");
    out.print(type);
    out.print(" ");
    out.print(get);
    out.print(propertyName);
    out.print("() {\n\t\ttry {\n\t\t\treturn ");
    out.print(Generators.generateCast(type, getSentence));
    out.print(";\n\t\t}\n\t\tcatch (org.openxava.converters.ConversionException ex) {\n\t\t\tex.printStackTrace();\n\t\t\tthrow new ");
    out.print(getException());
    out.print("(XavaResources.getString(\"generator.conversion_error\", \"");
    out.print(propertyName);
    out.print("\", \"");
    out.print(metaModel.getName());
    out.print("\", \"");
    out.print(type);
    out.print("\"));\n\t\t}\n\t}\n\t\n\t/**\n\t * ");
    out.print(ejbSetInterfaceMethod);
    out.print("\n\t */\n\t");
    out.print(propertyAccessLevel);
    out.print(" void set");
    out.print(propertyName);
    out.print("(");
    out.print(type);
    out.print(" new");
    out.print(propertyName);
    out.print(") {\n\t\ttry {");
    if (ejb) { 
    out.print(" \n\t\t\tthis.modified = true;");
    } 
    out.print(" \n\t\t\tset_");
    out.print(propertyName);
    out.print("(");
    out.print(Generators.generateCast(cmpType, setSentence));
    out.print(");\n\t\t}\n\t\tcatch (org.openxava.converters.ConversionException ex) {\n\t\t\tex.printStackTrace();\n\t\t\tthrow new ");
    out.print(getException());
    out.print("(XavaResources.getString(\"generator.conversion_error\", \"");
    out.print(propertyName);
    out.print("\", \"");
    out.print(metaModel.getName());
    out.print("\", \"");
    out.print(type);
    out.print("\"));\n\t\t}\t\t\n\t}");
    		
    		} else { // normal
    			if (!ejb) {
    			
    out.print(" \n\t");
    out.print(attributeAccessLevel);
    out.print(" ");
    out.print(type);
    out.print(" ");
    out.print(property.getName());
    out.print(";\n\t");
    out.print(propertyAccessLevel);
    out.print(" ");
    out.print(type);
    out.print(" ");
    out.print(get);
    out.print(propertyName);
    out.print("() {\n\t\treturn ");
    out.print(property.getName());
    out.print(";\n\t}\n\t");
    out.print(propertyAccessLevel);
    out.print(" void set");
    out.print(propertyName);
    out.print("(");
    out.print(type);
    out.print(" new");
    out.print(propertyName);
    out.print(") {\n\t\tthis.");
    out.print(property.getName());
    out.print(" = new");
    out.print(propertyName);
    out.print(";\n\t}");
    			
    			}
    			else {
    
    out.print(" \n\t/**\n\t * ");
    out.print(tagInterfaceMethod);
    out.print("\n\t * @ejb:persistent-field\n\t * ");
    out.print(ejbPkField);
    out.print("\n\t * ");
    out.print(tagValueObject);
    out.print("\n\t *\n\t * @jboss:column-name \"");
    out.print(column);
    out.print("\"\n\t */\n\tpublic abstract ");
    out.print(type);
    out.print(" ");
    out.print(get);
    out.print(propertyName);
    out.print("();\n\t/**\n\t  * ");
    out.print(ejbSetInterfaceMethod);
    out.print("\n\t  */\n\tpublic abstract void set");
    out.print(propertyName);
    out.print("(");
    out.print(type);
    out.print(" new");
    out.print(propertyName);
    out.print(");");
    }}} //else/else/else 
    
        } catch (Exception e) {
            System.out.println("Exception: "+e.getMessage());
            e.printStackTrace();
            return false;
        }
        return true;
    }
    
    /**
     * A program generator is typically (but not always) invoked
     * with a command line with arguments for the XML input file
     * and output file.
     */    
    public static void main(String[] args) {
        try {
            ProgramWriter out = args.length>=2
                ?new ProgramWriter(new FileOutputStream(args[1]))
                :new ProgramWriter(System.out);
            PropertyPG pg = new PropertyPG();
            for (int j=1; j<=args.length; ++j) {
                pg.properties.put("arg"+j, args[j-1]);
            }
            pg.generate(new XPathContext(args[0]), out);
            out.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
    
    /**
     * This array provides program generator development history
     */
    public String[][] history = {
        { "Fri Mar 21 14:15:30 CET 2014", // date this file was generated
             "../OpenXava/generator/property.xml", // input file
             "../OpenXava/generator/PropertyPG.java" }, // output file
        {"Mon Apr 09 16:45:30 EDT 2001", "TL2Java.xml", "TL2Java.java", }, 
        {"Mon Apr 09 16:39:37 EDT 2001", "TL2Java.xml", "TL2Java.java", }, 
        {"Mon Apr 09 16:37:21 EDT 2001", "TL2Java.xml", "TL2Java.java", }, 
        {"Fri Feb 09 14:49:11 EST 2001", "TL2Java.xml", "TL2Java.java", }, 
        {"Fri Feb 09 14:30:24 EST 2001", "TL2Java.xml", "TL2Java.java", }, 
        {"Fri Feb 09 11:13:01 EST 2001", "TL2Java.xml", "TL2Java.java", }, 
        {"Fri Feb 09 10:57:04 EST 2001", "TL2Java.xml", "TL2Java.java", }, 
        {"Wed Apr 26 11:15:41 EDT 2000", "..\\input\\TL2Java.xml", "TL2Java1.java", }, 
        {"April 2000", "hand coded", "TL2Java1.java", }, 

    };
}